### PR TITLE
feat: enable sequential ordering #244

### LINKED ISSUE #244: epic(donation): enable sequential ordering
## User Story

As an admin, I want sequential donation IDs so that I can satisfy the legal and accounting needs of my organization.

## Current Behavior

I currently have donation IDs based on the post ID which means there are large gaps between numbers that make it appear as if there might be missing donations from an accounting perspective.

## Expected Behavior

I expect to have sequential donation IDs without gaps between them.

European countries, and Canada require that transaction ID numbers be sequential. This isn't currently supported in Give. EDD and WooCommerce both did serious work to make that possible in both of their plugins.

### Resources

- [EDD Advanced Sequential Ordering Extension](https://easydigitaldownloads.com/downloads/advanced-sequential-order-numbers/)
- [Long Support Thread at EDD on their process](https://easydigitaldownloads.com/forums/topic/sequential-numbering/)
- [Ticket in EDD Github with links to their commits](https://github.com/easydigitaldownloads/Easy-Digital-Downloads/issues/2020)
- @chriscct7's [list for their approach to making it happen is very instructive](https://github.com/easydigitaldownloads/Easy-Digital-Downloads/issues/2020#issuecomment-35322424)
- [WooCommerce Sequential Ordering extension](https://wordpress.org/plugins/woocommerce-sequential-order-numbers/)
- https://www.canadiancharitylaw.ca/blog/cra_release_registered_charities_newsletter_no._33_-_information_on_receipt

### Support Tickets

- https://secure.helpscout.net/conversation/489032824/12578/?folderId=848135
- https://secure.helpscout.net/conversation/489032824/12578/?folderId=672199
- https://secure.helpscout.net/conversation/377605708/8176/?folderId=672199
- https://secure.helpscout.net/conversation/377605708/8176/?folderId=672199

### PR DESCRIPTION
**Pull request by ravinderk**
## Description
This PR will close #244 

## How Has This Been Tested?
- [x] Test with manually creating donations
- [x] Test donation detail page
- [x] Test Log setting page
- [x] Test table give_sequential_ordering  table
- [x]  Test Page title (serial code) on donations
- [x] Test receipt
- [x] Test history page
- [x] Test sequential ordering setting page
- [x] Test with Stripe and Razorpay
- [x] Test with pdf receipt addon https://github.com/WordImpress/Give-PDF-Receipts/tree/give210/sequential-donation-compatibility
- [x] Test PHPUnit

## Screenshots (jpeg or gifs if applicable):
2a2ta2zi9e
screen shot 2018-03-22 at 6 03 02 pm
image
image
image
image

## Types of changes
New feature (the non-breaking change which adds functionality)

## Noticeable changes
- Some unused deprecated function removed:
  - `give_format_payment_number `
  - `give_get_next_payment_number `
  - `give_remove_payment_prefix_postfix `
- Remove following unused setting keys:
  - `enable_sequential `
  - `sequential_prefix`
  - `sequential_postfix `
  - `give_last_payment_number`
- Sinve we are storingdonation number in `post_title` then ther are no need of custom serch logic: https://github.com/WordImpress/Give/pull/2936/files#diff-b594a31ba846d2a8e7d09d6c474364c0L597
- New table introduce `give_sequential_ordering` whic has two column `id` and `payment_id`
- Added new attribute `serial_number ` to `give_reciept` **(reverted. we are using `donation_id`)**
- Added new email tag `serial_number` **(reverted. we are using `payment_id`)**
- Updated PHPUnit test: https://github.com/WordImpress/Give/pull/2936/files#diff-31df4240052ebcab80a7d5491a2de60c
- Added function to decode time tags: `give_time_do_tags`
- `Give_Payment:number` will be set to serial number
- Add confrim before edit feature to `radio` and `text|number|email|password` type setting field
-  Donation history export will contain `Payment Number` column if feature enabled
## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

### CODE REVIEW

File: give.php [hunk 1]
```diff
@@ -229,6 +241,13 @@
 		 */
 		public $scripts;
 
+		/**
+		 * Giv_Seq_Donation_Number Object.
```
Comment (DevinWalker) [inline give.php]: Typo "Give_"

File: includes/class-give-db-sequential-donations.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+// Exit if accessed directly.
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * Give_DB_Sequential_Donations Class
+ *
+ * This class is for interacting with the log database table.
```
Comment (DevinWalker) [inline includes/class-give-db-sequential-donations.php]: Incorrect description

File: includes/misc-functions.php [hunk 3]
```diff
@@ -1761,4 +1761,47 @@ function give_maybe_define_constant( $name, $value ) {
 	if ( ! defined( $name ) ) {
 		define( $name, $value );
 	}
-}
\ No newline at end of file
+}
+
+/**
+ * Decode time short tage in string
```
Comment (DevinWalker) [inline includes/misc-functions.php]: Type "tag"

File: includes/misc-functions.php [hunk 4]
```diff
@@ -1761,4 +1761,47 @@ function give_maybe_define_constant( $name, $value ) {
...
+			date( 'Y', $current_time ),
+			date( 'Y', $current_time ),
+			date( 'G', $current_time ),
+			date( 'H', $current_time ),
+			date( 's', $current_time )
+		),
+		$string
+	);
+
+	return $string;
```
Comment (DevinWalker) [inline includes/misc-functions.php]: Filter needed

**Comment by DevinWalker** [comment]
## After Reviewing

1. The default starting number should be `1`. Currently it appears that there is no default "Sequential Starting Number": 



2. Missing field descriptions and unlock description. I assume you're working on this still? However, this PR was in the "Ready for QA" pipeline so if it's not ready please place it in the appropriate pipeline.





3. Ensure the Sequential Order feature is _enabled_ by default for _new_ installs.

4. Remove the word "Sequential" from field titles:



5. The first field should be "Sequential Ordering" and when toggled the subfields should show/hide:

2018-03-21_16-24-27 copy

**Comment by ravinderk** [comment]
@DevinWalker suggestion implemented.

**1. The default starting number should be 1. Currently, it appears that there is no default "Sequential Starting Number":** 
Ans: if the website already contains donations then starting number will be next to max donation id instead of `1`. On the fresh install, starting number will be `1`.

**Comment by DevinWalker** [comment]
@ravinderk when Sequential Ordering is enabled that number should display on the donation receipt. Currently it is displaying the post ID:

File: tests/unit-tests/tests-payments.php [hunk 5]
```diff
@@ -283,32 +283,92 @@ public function test_get_payment_completed_date_functions() {
...
-		$this->assertEquals( 2, give_get_next_payment_number() );
+		$payment = new Give_Payment( $payment_id );
+		$this->assertEquals( 1, $payment->number );
 
-		// Now disable sequential and ensure values come back as expected
-		$give_options['enable_sequential'] = 0;
+		/**
+		 * Case 2: enable sequential donation with prefix and suffix
+		 */
+		give_update_option( 'sequential-ordering_number_prefix', 'Give-' );
```
Comment (DevinWalker) [inline tests/unit-tests/tests-payments.php]: @ravinderk good work with adding unit tests proactively 👍

**Comment by DevinWalker** [comment]
Thanks @ravinderk - one additional change.

### Remove Post ID from Single Donation View

**Comment by ravinderk** [comment]
@DevinWalker Fixed.

**Pull request merged**